NPE on changing thing location

* OH2 Version: `2.3.0-1`
* Binding Version: `2.4.0.201809142025` (but also had this in `2.3.0`)

Problem: NPE on setting device location, PaperUI->Configuration->Things->edit thing->set "Location"

* Device name: Cooper RF9534-NSG
* Device version: ?
* Link to [database](http://www.cd-jackson.com/index.php/zwave/zwave-device-database/zwave-device-list): https://www.cd-jackson.com/index.php/zwave/zwave-device-database/zwave-device-list/devicesummary/467 (almost, `manufacturerRef` is `4449:0001` in my case)
* If the device is not in the database, provide a link to the manual: http://www.cooperindustries.com/content/dam/public/wiringdevices/products/documents/instruction_sheets/technology_rf9534_36_42_instruct.pdf

Stacktrace:

```
2018-09-15 21:45:50.616 [ERROR] [st.core.internal.thing.ThingResource] - Exception during HTTP PUT request for update config at 'things/zwave:device:512:node2/config'
java.lang.NullPointerException: null
        at org.openhab.binding.zwave.handler.ZWaveThingHandler.handleConfigurationUpdate(ZWaveThingHandler.java:805) [222:org.openhab.binding.zwave:2.4.0.201809142025]
        at org.eclipse.smarthome.core.thing.internal.ThingRegistryImpl.updateConfiguration(ThingRegistryImpl.java:92) [108:org.eclipse.smarthome.core.thing:0.10.0.oh230]
        at org.eclipse.smarthome.io.rest.core.internal.thing.ThingResource.updateConfiguration(ThingResource.java:425) [118:org.eclipse.smarthome.io.rest.core:0.10.0.oh230]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory$1.invoke(ResourceMethodInvocationHandlerFactory.java:81) [170:org.glassfish.jersey.
core.jersey-server:2.22.2]
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:144) [170:org.glassfish.jersey.core.j
ersey-server:2.22.2]
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:161) [170:org.glassfish.jersey.core.
jersey-server:2.22.2]
        at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:160) [170:org
.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:99) [170:org.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:389) [170:org.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:347) [170:org.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:102) [170:org.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.server.ServerRuntime$2.run(ServerRuntime.java:326) [170:org.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:271) [169:org.glassfish.jersey.core.jersey-common:2.22.2]
        at org.glassfish.jersey.internal.Errors$1.call(Errors.java:267) [169:org.glassfish.jersey.core.jersey-common:2.22.2]
        at org.glassfish.jersey.internal.Errors.process(Errors.java:315) [169:org.glassfish.jersey.core.jersey-common:2.22.2]
        at org.glassfish.jersey.internal.Errors.process(Errors.java:297) [169:org.glassfish.jersey.core.jersey-common:2.22.2]
        at org.glassfish.jersey.internal.Errors.process(Errors.java:267) [169:org.glassfish.jersey.core.jersey-common:2.22.2]
        at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:317) [169:org.glassfish.jersey.core.jersey-common:2.22.2]
        at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:305) [170:org.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:1154) [170:org.glassfish.jersey.core.jersey-server:2.22.2]
        at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:473) [167:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
        at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:427) [167:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:388) [167:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:341) [167:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
        at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:228) [167:org.glassfish.jersey.containers.jersey-container-servlet-core:2.22.2]
        at com.eclipsesource.jaxrs.publisher.internal.ServletContainerBridge.service(ServletContainerBridge.java:76) [15:com.eclipsesource.jaxrs.publisher:5.3.1.201602281253]
        at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:848) [85:org.eclipse.jetty.servlet:9.3.21.v20170918]
        at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:584) [85:org.eclipse.jetty.servlet:9.3.21.v20170918]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceServletHandler.doHandle(HttpServiceServletHandler.java:71) [183:org.ops4j.pax.web.pax-web-jetty:6.0.9]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:548) [82:org.eclipse.jetty.security:9.3.21.v20170918]
        at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:226) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1180) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.ops4j.pax.web.service.jetty.internal.HttpServiceContext.doHandle(HttpServiceContext.java:284) [183:org.ops4j.pax.web.pax-web-jetty:6.0.9]
        at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:512) [85:org.eclipse.jetty.servlet:9.3.21.v20170918]
        at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1112) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.ops4j.pax.web.service.jetty.internal.JettyServerHandlerCollection.handle(JettyServerHandlerCollection.java:80) [183:org.ops4j.pax.web.pax-web-jetty:6.0.9]
        at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:134) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.server.Server.handle(Server.java:534) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:333) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251) [84:org.eclipse.jetty.server:9.3.21.v20170918]
        at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:283) [76:org.eclipse.jetty.io:9.3.21.v20170918]
        at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:108) [76:org.eclipse.jetty.io:9.3.21.v20170918]
        at org.eclipse.jetty.io.SelectChannelEndPoint$2.run(SelectChannelEndPoint.java:93) [76:org.eclipse.jetty.io:9.3.21.v20170918]
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.executeProduceConsume(ExecuteProduceConsume.java:303) [87:org.eclipse.jetty.util:9.3.21.v20170918]
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.produceConsume(ExecuteProduceConsume.java:148) [87:org.eclipse.jetty.util:9.3.21.v20170918]
        at org.eclipse.jetty.util.thread.strategy.ExecuteProduceConsume.run(ExecuteProduceConsume.java:136) [87:org.eclipse.jetty.util:9.3.21.v20170918]
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:671) [87:org.eclipse.jetty.util:9.3.21.v20170918]
        at org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:589) [87:org.eclipse.jetty.util:9.3.21.v20170918]
        at java.lang.Thread.run(Thread.java:745) [?:?]
```

--- a/src/main/java/org/openhab/binding/zwave/handler/ZWaveThingHandler.java
+++ b/src/main/java/org/openhab/binding/zwave/handler/ZWaveThingHandler.java
@@ -567,8 +567,8 @@ public class ZWaveThingHandler extends ConfigStatusThingHandler implements ZWave
         Configuration configuration = editConfiguration();
         for (Entry<String, Object> configurationParameter : configurationParameters.entrySet()) {
             Object valueObject = configurationParameter.getValue();
-            logger.debug("NODE {}: Configuration update set {} to {}", nodeId, configurationParameter.getKey(),
-                    valueObject);
+            logger.debug("NODE {}: Configuration update set {} to {} ({})", nodeId, configurationParameter.getKey(),
+                    valueObject, valueObject == null ? "null" : valueObject.getClass().getSimpleName());
             String[] cfg = configurationParameter.getKey().split("_");
             switch (cfg[0]) {
                 case "config":
@@ -796,13 +796,19 @@ public class ZWaveThingHandler extends ConfigStatusThingHandler implements ZWave
                         continue;
                     }
 
+                    if (configurationParameter.getValue() == null
+                            || !(configurationParameter.getValue() instanceof String)) {
+                        logger.debug("NODE {}: Error setting NodeNamingCommandClass {} to invalid value {}", nodeId,
+                                cfg[1], configurationParameter.getValue(), configurationParameter.getValue());
+                        continue;
+                    }
+
                     if ("name".equals(cfg[1])) {
-                        controllerHandler.sendData(node.encapsulate(
-                                nameCommandClass.setNameMessage(configurationParameter.getValue().toString()), 0));
+                        node.sendMessage(nameCommandClass.setNameMessage(configurationParameter.getValue().toString()));
                     }
                     if ("location".equals(cfg[1])) {
-                        controllerHandler.sendData(node.encapsulate(
-                                nameCommandClass.setLocationMessage(configurationParameter.getValue().toString()), 0));
+                        node.sendMessage(
+                                nameCommandClass.setLocationMessage(configurationParameter.getValue().toString()));
                     }
                     pendingCfg.put(configurationParameter.getKey(), valueObject);
                     break;

--- a/src/test/java/org/openhab/binding/zwave/handler/ZWaveThingHandlerTest.java
+++ b/src/test/java/org/openhab/binding/zwave/handler/ZWaveThingHandlerTest.java
@@ -36,6 +36,7 @@ import org.openhab.binding.zwave.internal.protocol.ZWaveController;
 import org.openhab.binding.zwave.internal.protocol.ZWaveNode;
 import org.openhab.binding.zwave.internal.protocol.commandclass.ZWaveAssociationCommandClass;
 import org.openhab.binding.zwave.internal.protocol.commandclass.ZWaveCommandClass.CommandClass;
+import org.openhab.binding.zwave.internal.protocol.commandclass.ZWaveNodeNamingCommandClass;
 import org.openhab.binding.zwave.internal.protocol.commandclass.ZWaveWakeUpCommandClass;
 import org.openhab.binding.zwave.internal.protocol.transaction.ZWaveCommandClassTransactionPayload;
 
@@ -77,8 +78,8 @@ public class ZWaveThingHandlerTest {
         Field fieldControllerHandler;
         try {
             ZWaveWakeUpCommandClass wakeupClass = new ZWaveWakeUpCommandClass(node, controller, null);
-
             ZWaveAssociationCommandClass associationClass = new ZWaveAssociationCommandClass(node, controller, null);
+            ZWaveNodeNamingCommandClass namingClass = new ZWaveNodeNamingCommandClass(node, controller, null);
 
             Mockito.doNothing().when(node).sendMessage(payloadCaptor.capture());
             Mockito.doNothing().when(thingCallback).thingUpdated(Matchers.any(Thing.class));
@@ -95,6 +96,8 @@ public class ZWaveThingHandlerTest {
             Mockito.when(node.getCommandClass(Matchers.eq(CommandClass.COMMAND_CLASS_WAKE_UP))).thenReturn(wakeupClass);
             Mockito.when(node.getCommandClass(Matchers.eq(CommandClass.COMMAND_CLASS_ASSOCIATION)))
                     .thenReturn(associationClass);
+            Mockito.when(node.getCommandClass(Matchers.eq(CommandClass.COMMAND_CLASS_NODE_NAMING)))
+                    .thenReturn(namingClass);
         } catch (NoSuchFieldException | SecurityException e) {
             e.printStackTrace();
         } catch (IllegalArgumentException e) {
@@ -124,7 +127,31 @@ public class ZWaveThingHandlerTest {
     }
 
     @Test
-    public void TestConfigurationWakeup() {
+    public void testConfigurationLocation() {
+        ZWaveCommandClassTransactionPayload msg;
+        doConfigurationUpdate(ZWaveBindingConstants.CONFIGURATION_NODELOCATION, null);
+        doConfigurationUpdate(ZWaveBindingConstants.CONFIGURATION_NODELOCATION, Integer.valueOf(1));
+
+        List<ZWaveCommandClassTransactionPayload> response = doConfigurationUpdateCommands(
+                ZWaveBindingConstants.CONFIGURATION_NODELOCATION, "Testing");
+        assertEquals(1, response.size());
+        msg = response.get(0);
+        byte[] expectedResponse = { 0x77, 0x04, 0x00, 0x54, 0x65, 0x73, 0x74, 0x69, 0x6E, 0x67 };
+        assertTrue(Arrays.equals(msg.getPayloadBuffer(), expectedResponse));
+    }
+    
+    public String print(byte[] bytes) {
+    StringBuilder sb = new StringBuilder();
+    sb.append("[ ");
+    for (byte b : bytes) {
+        sb.append(String.format("0x%02X ", b));
+    }
+    sb.append("]");
+    return sb.toString();
+}
+
+    @Test
+    public void testConfigurationWakeup() {
         ZWaveCommandClassTransactionPayload msg;
         List<ZWaveCommandClassTransactionPayload> response = doConfigurationUpdateCommands(
                 ZWaveBindingConstants.CONFIGURATION_WAKEUPINTERVAL, new BigDecimal(600));
@@ -137,7 +164,7 @@ public class ZWaveThingHandlerTest {
     }
 
     @Test
-    public void TestConfigurationRepollPeriod() {
+    public void testConfigurationRepollPeriod() {
         ZWaveThingHandler handler = doConfigurationUpdate(ZWaveBindingConstants.CONFIGURATION_CMDREPOLLPERIOD,
                 new BigDecimal(600));
         assertEquals(1, handler.getThing().getConfiguration().getProperties().size());
@@ -156,7 +183,7 @@ public class ZWaveThingHandlerTest {
     }
 
     @Test
-    public void TestConfigurationAssociationList() {
+    public void testConfigurationAssociationList() {
         List<String> nodeList = new ArrayList<String>();
         nodeList.add("node_1");
         nodeList.add("node_2_1");
@@ -175,7 +202,7 @@ public class ZWaveThingHandlerTest {
     }
 
     @Test
-    public void TestConfigurationAssociationListController() {
+    public void testConfigurationAssociationListController() {
         List<String> nodeList = new ArrayList<String>();
         nodeList.add("controller");
         nodeList.add("node_2_1");
@@ -192,7 +219,7 @@ public class ZWaveThingHandlerTest {
     }
 
     @Test
-    public void TestConfigurationAssociationRoot() {
+    public void testConfigurationAssociationRoot() {
         List<ZWaveCommandClassTransactionPayload> response = doConfigurationUpdateCommands("group_1", "node_1");
 
         // Check that there are only 2 requests - the SET and GET
@@ -201,7 +228,7 @@ public class ZWaveThingHandlerTest {
     }
 
     @Test
-    public void TestConfigurationAssociationEndpoint() {
+    public void testConfigurationAssociationEndpoint() {
         List<ZWaveCommandClassTransactionPayload> response = doConfigurationUpdateCommands("group_1", "node_1_0");
 
         // Check that there are only 2 requests - the SET and GET

